src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,156 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (tail == null) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index, true);
+        if (index == size) {
+            add(value);
+        } else {
+            insertAtIndex(value, index);
+        }
+    }
+
+    private void insertAtIndex(T value, int index) {
+        Node<T> current = getNode(index);
+        Node<T> newNode = new Node<>(current.prev, value, current);
+        if (current.prev == null) {
+            head = newNode;
+        } else {
+            current.prev.next = newNode;
+        }
+        current.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, false);
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, false);
+        Node<T> node = getNode(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, false);
+        Node<T> nodeRemove = getNode(index);
+        return removeNode(nodeRemove);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if (Objects.equals(object, current.value)) {
+                removeNode(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private T removeNode(Node<T> node) {
+        final T value = node.value;
+        Node<T> nextNode = node.next;
+        Node<T> prevNode = node.prev;
+
+        if (prevNode == null) {
+            head = nextNode;
+        } else {
+            prevNode.next = nextNode;
+        }
+
+        if (nextNode == null) {
+            tail = prevNode;
+        } else {
+            nextNode.prev = prevNode;
+        }
+        node.prev = null;
+        node.next = null;
+        node.value = null;
+
+        size--;
+        return value;
+    }
```

Comment:
Create a separate method `unlink(Node node)`. The logic for removing a node is used multiple times, so it should be extracted into its own method to avoid duplication.